pennylane/devices/qubit_mixed/sampling.py

```diff
@@ -0,0 +1,311 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Submodule for sampling a qubit mixed state.
+"""
+# pylint: disable=too-many-positional-arguments, too-many-arguments
+import functools
+from typing import Callable
+
+import numpy as np
+
+import pennylane as qml
+from pennylane import math
+from pennylane.measurements import (
+    CountsMP,
+    ExpectationMP,
+    SampleMeasurement,
+    SampleMP,
+    Shots,
+    VarianceMP,
+)
+from pennylane.ops import Sum
+from pennylane.typing import TensorLike
+
+from .apply_operation import _get_num_wires, apply_operation
+from .measure import measure
+
+
+def _apply_diagonalizing_gates(
+    mp: SampleMeasurement, state: np.ndarray, is_state_batched: bool = False
+):
+    """Applies diagonalizing gates when necessary"""
+    if mp.obs:
+        for op in mp.diagonalizing_gates():
+            state = apply_operation(op, state, is_state_batched=is_state_batched)
+
+    return state
+
+
+def _process_samples(
+    mp,
+    samples,
+    wire_order,
+):
+    """Processes samples like SampleMP.process_samples, but different in need of some special cases e.g. CountsMP"""
+    wire_map = dict(zip(wire_order, range(len(wire_order))))
+    mapped_wires = [wire_map[w] for w in mp.wires]
+
+    if mapped_wires:
+        # if wires are provided, then we only return samples from those wires
+        samples = samples[..., mapped_wires]
+
+    num_wires = samples.shape[-1]  # wires is the last dimension
+
+    if mp.obs is None:
+        # if no observable was provided then return the raw samples
+        return samples
+
+    # Replace the basis state in the computational basis with the correct eigenvalue.
+    # Extract only the columns of the basis samples required based on ``wires``.
+    # This step converts e.g. 110 -> 6
+    powers_of_two = 2 ** qml.math.arange(num_wires)[::-1]  # e.g. [1, 2, 4, ...]
+    indices = qml.math.array(samples @ powers_of_two)
+    return mp.eigvals()[indices]
+
+
+def _process_counts_samples(processed_sample, mp_has_obs):
+    """Processes a set of samples and counts the results."""
+    observables, counts = math.unique(processed_sample, return_counts=True, axis=0)
```

Comment:
So `CountsMP` has this property `all_outcomes`. If `True`, it should return a count of 0 for an outcome that is not observed. Let's say `shots=5` and you get `1` from all 5 shots, then instead of returning `{1: 5}`, return `{1: 5, -1: 0}`. See `counts.py` for more details.